x/blob/README.md

```diff
@@ -2,44 +2,141 @@
+  // associated with this message should use when included in a block. The
+  // share_versions specified must match the share_versions used to generate the
+  // share_commitment in this message.
+  repeated uint32 share_versions = 8;
+}
+```
 
-When a block producer is preparing a block, they must perform an extra step for `BlobTx`s so that end-users can find the blob shares relevant to their submitted `BlobTx`. In particular, block proposers wrap the `BlobTx` in the PayForBlobs namespace with the index of the first share of the blob in the data square. See [Non-interactive Default Rules](https://github.com/celestiaorg/celestia-specs/blob/master/src/rationale/message_block_layout.md#non-interactive-default-rules) for more details.
+### Generating the `ShareCommitment`
 
-Since `BlobTx`s can contain multiple blobs, the `BlobTx` is wrapped with one share index per blob in the transaction. The index wrapped transaction is called an [IndexWrapper](https://github.com/celestiaorg/celestia-core/blob/2d2a65f59eabf1993804168414b86d758f30c383/proto/tendermint/types/types.proto#L192-L198) and this is the type that gets marshalled and written to the PayForBlobNamespace.
+1. Split the blob into shares of size `appconsts.ShareSize`
+1. Determine the
+   [`SubtreeWidth`](https://github.com/celestiaorg/celestia-app/blob/v1.0.0-rc2/pkg/shares/non_interactive_defaults.go#L94-L116)
+   by dividing the length in shares by the `SubtreeRootThreshold`.
+1. Generate each subtree root by diving the blob shares into `SubtreeWidth`
+   sized sets, then take the binary [namespaced merkle tree
+   (NMT)](https://github.com/celestiaorg/nmt/blob/v0.16.0/docs/spec/nmt.md) root
+   of each set of shares.
+1. Calculate the final share commitment by taking the merkle root (note: not an
+   NMT, just a normal binary merkle root) of the subtree roots from the previous
+   step.
+
+See
+[`CreateCommitment`](https://github.com/celestiaorg/celestia-app/blob/v1.0.0-rc2/x/blob/types/payforblob.go#L169-L236)
+for an implementation. See [data square
+layout](../../specs/src/specs/data_square_layout.md) and
+[ADR013](../../docs/architecture/adr-013-non-interactive-default-rules-for-zero-padding.md)
+for details on the rational of the square layout.
+
+## Validity Rules
+
+In order for a proposal block to be considered valid, each `BlobTx`, and thus
+each PFB, to be included in a block must follow a set of validity rules.
+
+1. Signatures: All blob transactions must have valid signatures. This is
+   state-dependent because correct signatures require using the correct sequence
+   number(aka nonce).
+1. Single SDK.Msg: There must be only a single sdk.Msg in the blob transaction.
+1. Namespace Validity: The namespace of each blob transaction must be valid.
+   This validity is determined by the following sub-rules:
+    1. The namespace is not within the reserved namespace range.
+    1. The namespace is not the tail padding or parity namespaces.
```

Comment:
Would be great to add links for each different reserved namespace mentioned in these two lines.